gitlab/resource_gitlab_project_approval_rule.go

```diff
@@ -0,0 +1,429 @@
+		},
+	}
+}
+
+func resourceGitlabProjectApprovalRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	options := gitlab.CreateProjectLevelRuleOptions{
+		Name:              gitlab.String(d.Get("name").(string)),
+		ApprovalsRequired: gitlab.Int(d.Get("approvals_required").(int)),
+		UserIDs:           expandApproverIds(d.GetOk("user_ids")),
+		GroupIDs:          expandApproverIds(d.GetOk("group_ids")),
+	}
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab project-level rule %+v", project, options)
+
+	client := meta.(*gitlab.Client)
+	if rule, _, err := client.Projects.CreateProjectApprovalRule(project, &options); err == nil {
+		d.SetId(createApprovalRuleID(rule.ID, project))
+
+		return resourceGitlabProjectApprovalRuleRead(d, meta)
+	}
+
+	return err
+}
+
+func resourceGitlabProjectApprovalRuleRead(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	log.Printf("[DEBUG] read gitlab project-level rule %s", d.Id())
+
+	d.Partial(true)
+
+	projectID, _, err := splitApprovalRuleID(d.Id())
+	if err != nil {
+		return err
+	}
+	d.Set("project", projectID)
+	d.SetPartial("project")
+
+	rule, err := getApprovalRuleByID(meta, d.Id())
+	if err != nil || rule == nil {
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("unable to read GitLab approvel rule %s", d.Id())
+	}
+
+	d.Set("name", rule.Name)
+	d.SetPartial("name")
+	d.Set("approvals_required", rule.ApprovalsRequired)
+	d.SetPartial("approvals_required")
+
+	groupIDs := flattenApprovalRuleGroupIDs(rule.Groups)
+	if err := d.Set("group_ids", groupIDs); err != nil {
+		return err
+	}
+	d.SetPartial("group_ids")
+
+	eligibleApprovers := flattenApprovalRuleUserIDs(rule.EligibleApprovers)
+	if err := d.Set("eligible_approvers", eligibleApprovers); err != nil {
+		return err
+	}
+	d.SetPartial("eligible_approvers")
+
+	userIDs, err := getProjectApprovalRuleUserIDs(meta.(*gitlab.Client), d.Id())
+	if err != nil {
+		return err
+	}
+
+	if err := d.Set("user_ids", userIDs); err != nil {
+		return err
+	}
+
+	d.Partial(false)
+
+	return nil
+}
+
+func resourceGitlabProjectApprovalRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	if project, ruleID, err := splitApprovalRuleID(d.Id()); err == nil {
+		options := gitlab.UpdateProjectLevelRuleOptions{
+			Name:              gitlab.String(d.Get("name").(string)),
+			ApprovalsRequired: gitlab.Int(d.Get("approvals_required").(int)),
+			UserIDs:           expandApproverIds(d.GetOk("user_ids")),
+			GroupIDs:          expandApproverIds(d.GetOk("group_ids")),
+		}
+
+		ruleIDInt, err := strconv.Atoi(ruleID)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("[DEBUG] Project %s update gitlab project-level approval rule %s", project, *options.Name)
+
+		client := meta.(*gitlab.Client)
+		if _, _, err = client.Projects.UpdateProjectApprovalRule(project, ruleIDInt, &options); err == nil {
+			return resourceGitlabProjectVariableRead(d, meta)
```

Comment:
`resourceGitlabProjectVariableRead`? Shouldn't it be `resourceGitlabProjectApprovalRuleRead`?

Reply:
^ Yes please fix, if you need to even make this additional API call here. Though, you should already have all the data you need in the return value from your call to `UpdateProjectApprovalRule`, so I would prefer if you use that return object than to make another API call, to reduce the number of calls and to mitigate potential data consistency problems.

Reply:
Never mind. I see it is a recommended practice in the docs. Carry on!